render/tests/test-component-class.js

```diff
@@ -0,0 +1,702 @@
+					o(vnode.dom).equals(root.firstChild)
+					o(root.childNodes.length).equals(1)
+				},
+				view: function() {
+					return [{tag: "div", attrs: {id: "a"}, text: "b"}]
+				}
+			})
+
+			render(root, [{tag: component}])
+
+			o(called).equals(0)
+
+			render(root, [{tag: component}])
+
+			o(called).equals(1)
+			o(root.firstChild.nodeName).equals("DIV")
+			o(root.firstChild.attributes["id"].nodeValue).equals("a")
+			o(root.firstChild.firstChild.nodeValue).equals("b")
+		})
+		o("calls onremove", function() {
+			var called = 0
+			var component = create({
+				onremove: function(vnode) {
+					called++
+
+					o(Object.getPrototypeOf(this)).equals(component.prototype)
+					o(vnode.state).equals(this)
+					o(vnode.dom).equals(root.firstChild)
+					o(root.childNodes.length).equals(1)
+				},
+				view: function() {
+					return {tag: "div", attrs: {id: "a"}, text: "b"}
+				}
+			})
+
+			render(root, [{tag: component}])
+
+			o(called).equals(0)
+
+			render(root, [])
+
+			o(called).equals(1)
+			o(root.childNodes.length).equals(0)
+		})
+		o("calls onremove when returning fragment", function() {
+			var called = 0
+			var component = create({
+				onremove: function(vnode) {
+					called++
+
+					o(Object.getPrototypeOf(this)).equals(component.prototype)
+					o(vnode.state).equals(this)
+					o(vnode.dom).equals(root.firstChild)
+					o(root.childNodes.length).equals(1)
+				},
+				view: function() {
+					return [{tag: "div", attrs: {id: "a"}, text: "b"}]
+				}
+			})
+
+			render(root, [{tag: component}])
+
+			o(called).equals(0)
+
+			render(root, [])
+
+			o(called).equals(1)
+			o(root.childNodes.length).equals(0)
+		})
+		o("calls onbeforeremove", function() {
+			var called = 0
+			var component = create({
+				onbeforeremove: function(vnode, done) {
+					called++
+
+					o(Object.getPrototypeOf(this)).equals(component.prototype)
+					o(vnode.state).equals(this)
+					o(vnode.dom).equals(root.firstChild)
+					o(root.childNodes.length).equals(1)
+
+					done()
+				},
+				view: function() {
+					return {tag: "div", attrs: {id: "a"}, text: "b"}
+				}
+			})
+
+			render(root, [{tag: component}])
+
+			o(called).equals(0)
+
+			render(root, [])
+
+			o(called).equals(1)
+			o(root.childNodes.length).equals(0)
+		})
+		o("calls onbeforeremove when returning fragment", function() {
+			var called = 0
+			var component = create({
+				onbeforeremove: function(vnode, done) {
+					called++
+
+					o(Object.getPrototypeOf(this)).equals(component.prototype)
+					o(vnode.state).equals(this)
+					o(vnode.dom).equals(root.firstChild)
+					o(root.childNodes.length).equals(1)
+
+					done()
+				},
+				view: function() {
+					return [{tag: "div", attrs: {id: "a"}, text: "b"}]
+				}
+			})
+
+			render(root, [{tag: component}])
+
+			o(called).equals(0)
+
+			render(root, [])
+
+			o(called).equals(1)
+			o(root.childNodes.length).equals(0)
+		})
+		o("does not recycle when there's an onupdate", function() {
```

Comment:
This is the only failing test, BTW. It's odd that the corresponding object test is passing, but this is failing, since they *should* be on identical code paths.

This is where I need a little help figuring out what *could* be causing this.